ssh-pkcs11-client.c

```diff
@@ -286,6 +446,37 @@ pkcs11_start_helper(void)
 		error("socketpair: %s", strerror(errno));
 		return (-1);
 	}
+#ifdef WINDOWS
+	if (posix_spawn_file_actions_init(&actions) != 0) {
+		error_f("posix_spawn_file_actions_init failed");
+		goto out;
+	}
+	actions_inited = 1;
+	if (posix_spawn_file_actions_adddup2(&actions, pair[1],
+		STDIN_FILENO) != 0 ||
+		posix_spawn_file_actions_adddup2(&actions, pair[1],
+			STDOUT_FILENO) != 0) {
+		error_f("posix_spawn_file_actions_adddup2 failed");
+		goto out;
+	}
+
+	av[0] = helper;
+	av[1] = verbosity;
+	av[2] = NULL;
+
+	if ((client_process_handle = OpenProcess(PROCESS_QUERY_INFORMATION, FALSE, sshagent_client_pid)) == NULL ||
+		OpenProcessToken(client_process_handle, TOKEN_QUERY | TOKEN_ASSIGN_PRIMARY, &client_token) == FALSE) {
```

Comment:
We shouldn't use the same token, we should duplicate the token and pass it to line 473.
I will take care of this. I will merge your PR.
Thanks a lot for your contributions.

Reply:
One more issue, over SSH to localhost this wouldn't work. Because the SSH session is in session 0, smart card authentication wouldn't work in session 0. I think it's fine because customer is expected to physically present with the smart card and there wouldn't be a need to do over SSH to localhost. .

Reply:
@bagajjal , Can you be a bit more clear about the scenario you are describing, please?
The helper shouldn't run in session 0 since this session has elevated privileges and the whole point here is for the helper to be executed with the same (reduced) privileges as the client (which doesn't run in session 0).
This implementation archives that and `ssh` to localhost works as shown in the screenshot below
![testing-ssh-pkcs11-helper](https://user-images.githubusercontent.com/1812956/154601505-c0dbd5be-00c6-490f-87dd-d366809e06aa.png)

Reply:
That's the normal behavior. I'm talking about SSH connection made over the SSH.
First you make an SSH connection to the machine (which has smart card). Now over the SSH connection, if you try to run "ssh-add -s <pkcs11-provider>" then it will fail because the ssh-add.exe process id is running in session 0. 

As I said, this is a corner case as we expect the customer to be physically present with the smart card to use it. When customer is physically present, there is no point why customer should try ssh-add.exe over an SSH connection.

Reply:
@bagajjal , What you are describing requires [ssh-agent forwarding](https://docs.github.com/en/developers/overview/using-ssh-agent-forwarding) to be enabled.
On a Linux-based system, with agent-forwarding enabled this scenario is supported and would work.
This is one of the reasons Yubikeys have the [capability to set touch policy](https://docs.yubico.com/yesdk/users-manual/application-piv/pin-touch-policies.html#touch-policies) on pkcs11 keys - If an authentication attempt was made through a forwarded agent on a remote host, the user will be made aware of this (the Yubikey will flash) and will have to physically touch the key for the authentication to proceed.
If agent forwarding is implemented for Windows, this should also work. I'll try to test this out soon.

Reply:
@bagajjal , What you are describing requires [ssh-agent forwarding](https://docs.github.com/en/developers/overview/using-ssh-agent-forwarding) to be enabled  (except, running `ssh-add -s` on the jump host doesn't make much sense. Running `ssh <target_host> with the forwarded agent does`).
On a Linux-based system, with agent-forwarding enabled the following scenario is supported and would work:
1) Client machine initiates ssh connection to the jump host with agent forwarding using a pkcs11 key.
2) Jump host machine initiates ssh connection to the target host - since the _agent was forwarded_ from the client machine via a socket- the authentication (sign request) will happen on the client machine, where the smartcard is connected.
This is one of the reasons Yubikeys have the [capability to set touch policy](https://docs.yubico.com/yesdk/users-manual/application-piv/pin-touch-policies.html#touch-policies) on pkcs11 keys - If an authentication attempt was made through a forwarded agent on a remote host (where the Yubikey isn't connected), the user will be made aware of this (the Yubikey will flash) and will have to physically touch the key for the authentication to proceed.
If agent forwarding is implemented for Windows, this should also work. I'll try to test this out soon.

Reply:
Agree. Windows is missing ssh-agent forwarding. It's there in our backlog. It would be great if you can contribute for the ssh-agent forwarding.